# Analyzing The Chinook Data Set 

In this project I am working with a data set of an online music store. I will use MySQL for my analysis. There are different tables in this database, which i will combine to find the answers I am looking for. Let's take a look at the structure of the database first.

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

In [2]:
%%sql
SELECT * FROM sqlite_master

 * sqlite:///chinook.db
Done.


[('table', 'album', 'album', 2, 'CREATE TABLE [album]\n(\n    [album_id] INTEGER PRIMARY KEY NOT NULL,\n    [title] NVARCHAR(160)  NOT NULL,\n    [artist_id] INTEGER  NOT NULL,\n    FOREIGN KEY ([artist_id]) REFERENCES [artist] ([artist_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'artist', 'artist', 3, 'CREATE TABLE [artist]\n(\n    [artist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'customer', 'customer', 4, 'CREATE TABLE [customer]\n(\n    [customer_id] INTEGER PRIMARY KEY NOT NULL,\n    [first_name] NVARCHAR(40)  NOT NULL,\n    [last_name] NVARCHAR(20)   ... (276 characters truncated) ...  [support_rep_id] INTEGER,\n    FOREIGN KEY ([support_rep_id]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'employee', 'employee', 5, 'CREATE TABLE [employee]\n(\n    [employee_id] INTEGER PRIMARY KEY NOT NULL,\n    [last_name] NVARCHAR(20)  NOT NULL,\n    [first_name] NVARCHAR(20)   ... (310 characters truncated) ... 4),\n    [email] NVARCHAR(60),\n    FOREIGN KEY ([reports_to]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'genre', 'genre', 6, 'CREATE TABLE [genre]\n(\n    [genre_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'invoice', 'invoice', 7, 'CREATE TABLE [invoice]\n(\n    [invoice_id] INTEGER PRIMARY KEY NOT NULL,\n    [customer_id] INTEGER  NOT NULL,\n    [invoice_date] DATETIME  NOT NUL ... (195 characters truncated) ... tal] NUMERIC(10,2)  NOT NULL,\n    FOREIGN KEY ([customer_id]) REFERENCES [customer] ([customer_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'invoice_line', 'invoice_line', 8, 'CREATE TABLE [invoice_line]\n(\n    [invoice_line_id] INTEGER PRIMARY KEY NOT NULL,\n    [invoice_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT ... (161 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'media_type', 'media_type', 9, 'CREATE TABLE [media_type]\n(\n    [media_type_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist', 'playlist', 10, 'CREATE TABLE [playlist]\n(\n    [playlist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist_track', 'playlist_track', 11, 'CREATE TABLE [playlist_track]\n(\n    [playlist_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT NULL,\n    CONSTRAINT [pk_playlist_track] PRIMARY ... (114 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('index', 'sqlite_autoindex_playlist_track_1', 'playlist_track', 12, None),
 ('table', 'track', 'track', 13, 'CREATE TABLE [track]\n(\n    [track_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(200)  NOT NULL,\n    [album_id] INTEGER,\n    [media_type_ ... (377 characters truncated) ... ON ON UPDATE NO ACTION,\n    FOREIGN KEY ([media_type_id]) REFERENCES [media_type] ([media_type_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('index', 'ifk_albumartist_id', 'album', 16, 'CREATE INDEX [ifk_albumartist_id] ON [album] ([artist_id])'),
 ('index', 'ifk_customersupport_rep_id', 'customer', 17, 'CREATE INDEX [ifk_customersupport_rep_id] ON [customer] ([support_rep_id])'),
 ('index', 'ifk_employeereports_to', 'employee', 18, 'CREATE INDEX [ifk_employeereports_to] ON [employee] ([reports_to])'),
 ('index', 'ifk_invoicecustomer_id', 'invoice', 19, 'CREATE INDEX [ifk_invoicecustomer_id] ON [invoice] ([customer_id])'),
 ('index', 'ifk_invoice_lineinvoice_id', 'invoice_line', 20, 'CREATE INDEX [ifk_invoice_lineinvoice_id] ON [invoice_line] ([invoice_id])'),
 ('index', 'ifk_invoice_linetrack_id', 'invoice_line', 21, 'CREATE INDEX [ifk_invoice_linetrack_id] ON [invoice_line] ([track_id])'),
 ('index',

### Most Popular Genres in the USA

The first question I want to answer is: Which are the most popular genres in the USA?
To answer this question, I am writing a query that returns each genre, with the number of tracks sold in the USA, both in absolute numbers and percentages.

In [3]:
%%sql
WITH apc AS 
(SELECT il.*, i.billing_country FROM invoice_line il 
INNER JOIN invoice i ON i.invoice_id = il.invoice_id
 WHERE i.billing_country = 'USA'
 GROUP BY il.invoice_id)


SELECT g.name, 
ROUND(SUM(apc.quantity * apc.unit_price), 2) amount, 
ROUND(SUM(apc.unit_price*apc.quantity) / (SELECT SUM(unit_price*quantity) FROM apc)*100, 2) percentage, 
apc.billing_country 
FROM apc
INNER JOIN track t ON t.track_id = apc.track_id
INNER JOIN genre g ON t.genre_id = g.genre_id
GROUP BY g.name ORDER BY amount DESC





 * sqlite:///chinook.db
Done.


[('Rock', 71.28, 54.96, 'USA'),
 ('Alternative & Punk', 16.83, 12.98, 'USA'),
 ('Metal', 9.9, 7.63, 'USA'),
 ('R&B/Soul', 6.93, 5.34, 'USA'),
 ('Latin', 4.95, 3.82, 'USA'),
 ('Pop', 3.96, 3.05, 'USA'),
 ('Easy Listening', 2.97, 2.29, 'USA'),
 ('Classical', 2.97, 2.29, 'USA'),
 ('Blues', 2.97, 2.29, 'USA'),
 ('Jazz', 1.98, 1.53, 'USA'),
 ('Alternative', 1.98, 1.53, 'USA'),
 ('Reggae', 0.99, 0.76, 'USA'),
 ('Hip Hop/Rap', 0.99, 0.76, 'USA'),
 ('Electronica/Dance', 0.99, 0.76, 'USA')]

The TOP 3 most popular genres for the american customer base in the Chinook online store are Rock (51 %), Metal (14 %) and Alternative & Punk (14 %).

### Analyzing Sales Agent Performance

Next, i want to analyze the performance of sales agents. I will write a query, that assigns a total dollar amount to each employee within the company.

In [4]:
%%sql 
SELECT e.first_name || " " || e.last_name name, e.title, ROUND(SUM(i.total), 2) "total amount"
FROM employee e
LEFT JOIN customer c ON e.employee_id = c.support_rep_id
LEFT JOIN invoice i ON i.customer_id = c.customer_id
GROUP BY name ORDER BY "total amount" DESC


 * sqlite:///chinook.db
Done.


[('Jane Peacock', 'Sales Support Agent', 1731.51),
 ('Margaret Park', 'Sales Support Agent', 1584.0),
 ('Steve Johnson', 'Sales Support Agent', 1393.92),
 ('Robert King', 'IT Staff', None),
 ('Nancy Edwards', 'Sales Manager', None),
 ('Michael Mitchell', 'IT Manager', None),
 ('Laura Callahan', 'IT Staff', None),
 ('Andrew Adams', 'General Manager', None)]

### Market Analysis by Country

For the next step I want to analyze the different national markets. To achieve that, I group by country and calculate the total number of customers, the total and average sale value aswell as the average number of orders per customers.

To make these numbers easier to understand, I want to group countries with only one customer under the umbrella of "Other". I use a subquery with a CASE condition to achieve this. 

The result should be odered by total value of sales but I don't want "Other" to be on top of the list, since it is not a real country. I solve this by creating a hidden category, that I primarily order by while ordering by total value of sales behind that.

In [5]:
%%sql

WITH country_groups AS 
(SELECT country, 
CASE WHEN COUNT(customer_id) = 1 THEN 'Other' ELSE country END as country_or_other, 
CASE WHEN COUNT(customer_id) = 1 THEN 1 ELSE 0 END as country_group,
COUNT(customer_id) as total_customers_per_country 
FROM customer GROUP BY country ORDER BY 3)

SELECT 
cg.country_or_other country_name, 
COUNT (DISTINCT c.customer_id) total_customers, 
ROUND(SUM(i.total), 2) total_sales_value, 
ROUND(SUM(i.total) / COUNT(DISTINCT c.customer_id), 2) avg_sales_per_customer, 
ROUND(SUM(i.total) / COUNT(i.invoice_id), 2) avg_order_value
FROM customer c 
LEFT JOIN invoice i ON c.customer_id = i.customer_id
LEFT JOIN country_groups cg ON c.country = cg.country
GROUP BY 1 ORDER BY cg.country_group, 2 DESC

 * sqlite:///chinook.db
Done.


[('USA', 13, 1040.49, 80.04, 7.94),
 ('Canada', 8, 535.59, 66.95, 7.05),
 ('Brazil', 5, 427.68, 85.54, 7.01),
 ('France', 5, 389.07, 77.81, 7.78),
 ('Germany', 4, 334.62, 83.66, 8.16),
 ('United Kingdom', 3, 245.52, 81.84, 8.77),
 ('Czech Republic', 2, 273.24, 136.62, 9.11),
 ('India', 2, 183.15, 91.57, 8.72),
 ('Portugal', 2, 185.13, 92.57, 6.38),
 ('Other', 15, 1094.94, 73.0, 7.45)]

### Single Track Purchases vs. Album Purchases

I am interested in our customers behaviour regarding their purchasing of full albums. The background is, that the store is considering not to buy full albums anymore, if the customers are only interested in the more popular tracks and not the complete album. This would be a way to decrease the cost of the store. However, if there is a significant group of customers, who like to buy full albums, they might be dissatisified in the future.
Because there is no category for album purchase, we need to find a way to gather this information.

I wrote a query that produces a list of all unique albums first, including the number of individual tracks. Next, I wrote a query that groups by invoide_id and album_id. This basically assigns the purchases to the albums they belong to.

In [6]:
%%sql 
SELECT DISTINCT(t.album_id), a.title, COUNT(t.track_id)
FROM track t
LEFT JOIN album a ON a.album_id = t.album_id
GROUP BY a.album_id

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock We Salute You', 10),
 (2, 'Balls to the Wall', 1),
 (3, 'Restless and Wild', 3),
 (4, 'Let There Be Rock', 8),
 (5, 'Big Ones', 15),
 (6, 'Jagged Little Pill', 13),
 (7, 'Facelift', 12),
 (8, 'Warner 25 Anos', 14),
 (9, 'Plays Metallica By Four Cellos', 8),
 (10, 'Audioslave', 14),
 (11, 'Out Of Exile', 12),
 (12, 'BackBeat Soundtrack', 12),
 (13, 'The Best Of Billy Cobham', 8),
 (14, 'Alcohol Fueled Brewtality Live! [Disc 1]', 13),
 (15, 'Alcohol Fueled Brewtality Live! [Disc 2]', 5),
 (16, 'Black Sabbath', 7),
 (17, 'Black Sabbath Vol. 4 (Remaster)', 10),
 (18, 'Body Count', 17),
 (19, 'Chemical Wedding', 11),
 (20, 'The Best Of Buddy Guy - The Millenium Collection', 11),
 (21, 'Prenda Minha', 18),
 (22, 'Sozinho Remix Ao Vivo', 3),
 (23, 'Minha Historia', 34),
 (24, 'Afrociberdelia', 23),
 (25, 'Da Lama Ao Caos', 13),
 (26, 'Acústico MTV [live]', 17),
 (27, 'Cidade Negra - Hits', 14),
 (28, 'Na Pista', 10),
 (29, 'Axé Bahia 2001', 14),
 (30, 'BBC Sessions [Disc 1] [live]', 14),
 (31, 'Bongo Fury', 9),
 (32, 'Carnaval 2001', 14),
 (33, 'Chill: Brazil (Disc 1)', 17),
 (34, 'Chill: Brazil (Disc 2)', 17),
 (35, 'Garage Inc. (Disc 1)', 11),
 (36, 'Greatest Hits II', 17),
 (37, 'Greatest Kiss', 20),
 (38, 'Heart of the Night', 12),
 (39, 'International Superhits', 21),
 (40, 'Into The Light', 12),
 (41, 'Meus Momentos', 14),
 (42, 'Minha História', 14),
 (43, 'MK III The Final Concerts [Disc 1]', 7),
 (44, 'Physical Graffiti [Disc 1]', 6),
 (45, 'Sambas De Enredo 2001', 14),
 (46, 'Supernatural', 13),
 (47, 'The Best of Ed Motta', 14),
 (48, 'The Essential Miles Davis [Disc 1]', 13),
 (49, 'The Essential Miles Davis [Disc 2]', 10),
 (50, 'The Final Concerts (Disc 2)', 4),
 (51, "Up An' Atom", 22),
 (52, 'Vinícius De Moraes - Sem Limite', 15),
 (53, 'Vozes do MPB', 14),
 (54, 'Chronicle, Vol. 1', 20),
 (55, 'Chronicle, Vol. 2', 20),
 (56, 'Cássia Eller - Coleção Sem Limite [Disc 2]', 15),
 (57, 'Cássia Eller - Sem Limite [Disc 1]', 15),
 (58, 'Come Taste The Band', 9),
 (59, 'Deep Purple In Rock', 7),
 (60, 'Fireball', 7),
 (61, "Knocking at Your Back Door: The Best Of Deep Purple in the 80's", 11),
 (62, 'Machine Head', 7),
 (63, 'Purpendicular', 12),
 (64, 'Slaves And Masters', 9),
 (65, 'Stormbringer', 9),
 (66, 'The Battle Rages On', 10),
 (67, "Vault: Def Leppard's Greatest Hits", 16),
 (68, 'Outbreak', 9),
 (69, 'Djavan Ao Vivo - Vol. 02', 13),
 (70, 'Djavan Ao Vivo - Vol. 1', 13),
 (71, 'Elis Regina-Minha História', 14),
 (72, 'The Cream Of Clapton', 18),
 (73, 'Unplugged', 30),
 (74, 'Album Of The Year', 12),
 (75, 'Angel Dust', 14),
 (76, 'King For A Day Fool For A Lifetime', 15),
 (77, 'The Real Thing', 11),
 (78, 'Deixa Entrar', 14),
 (79, 'In Your Honor [Disc 1]', 10),
 (80, 'In Your Honor [Disc 2]', 10),
 (81, 'One By One', 11),
 (82, 'The Colour And The Shape', 13),
 (83, 'My Way: The Best Of Frank Sinatra [Disc 1]', 24),
 (84, 'Roda De Funk', 16),
 (85, 'As Canções de Eu Tu Eles', 14),
 (86, 'Quanta Gente Veio Ver (Live)', 15),
 (87, 'Quanta Gente Veio ver--Bônus De Carnaval', 3),
 (88, 'Faceless', 12),
 (89, 'American Idiot', 13),
 (90, 'Appetite for Destruction', 12),
 (91, 'Use Your Illusion I', 16),
 (92, 'Use Your Illusion II', 14),
 (93, 'Blue Moods', 13),
 (94, 'A Matter of Life and Death', 11),
 (95, 'A Real Dead One', 12),
 (96, 'A Real Live One', 11),
 (97, 'Brave New World', 10),
 (98, 'Dance Of Death', 11),
 (99, 'Fear Of The Dark', 12),
 (100, 'Iron Maiden', 9),
 (101, 'Killers', 10),
 (102, 'Live After Death', 18),
 (103, 'Live At Donington 1992 (Disc 1)', 10),
 (104, 'Live At Donington 1992 (Disc 2)', 10),
 (105, 'No Prayer For The Dying', 10),
 (106, 'Piece Of Mind', 9),
 (107, 'Powerslave', 8),
 (108, 'Rock In Rio [cd1]', 10),
 (109, 'Rock In Rio [cd2]', 9),
 (110, 'Seventh Son of a Seventh Son', 8),
 (111, 'Somewhere in Time', 8),
 (112, 'The Number of The Beast', 8),
 (113, 'The X Factor', 11),
 (114, 'Virtual XI', 8),
 (115, 'Sex Machine', 20),
 (116, 'Emergency On P

In [7]:
%%sql 
SELECT il.invoice_id, COUNT(t.track_id) number_tracks, t.album_id 
FROM invoice_line il 
INNER JOIN track t on t.track_id = il.track_id
GROUP BY il.invoice_id, t.album_id

 * sqlite:///chinook.db
Done.


[(1, 16, 91),
 (2, 1, 20),
 (2, 1, 34),
 (2, 1, 39),
 (2, 1, 66),
 (2, 1, 73),
 (2, 1, 134),
 (2, 1, 190),
 (2, 1, 215),
 (2, 1, 218),
 (2, 1, 322),
 (3, 1, 203),
 (3, 1, 214),
 (4, 1, 58),
 (4, 1, 67),
 (4, 1, 146),
 (4, 1, 153),
 (4, 1, 203),
 (4, 1, 207),
 (4, 1, 260),
 (4, 1, 314),
 (5, 17, 163),
 (6, 1, 5),
 (6, 1, 83),
 (7, 1, 6),
 (7, 1, 55),
 (7, 1, 67),
 (7, 1, 87),
 (7, 1, 106),
 (7, 1, 163),
 (7, 1, 211),
 (7, 1, 214),
 (7, 1, 239),
 (7, 1, 305),
 (7, 1, 322),
 (8, 1, 9),
 (8, 1, 46),
 (8, 1, 118),
 (8, 1, 125),
 (8, 1, 144),
 (8, 1, 145),
 (8, 1, 183),
 (8, 1, 185),
 (8, 1, 203),
 (8, 1, 233),
 (9, 1, 20),
 (9, 1, 21),
 (9, 1, 79),
 (9, 1, 83),
 (9, 1, 101),
 (9, 1, 162),
 (9, 1, 173),
 (9, 1, 271),
 (9, 1, 321),
 (10, 1, 215),
 (10, 1, 257),
 (11, 1, 49),
 (11, 1, 89),
 (11, 1, 115),
 (11, 1, 181),
 (11, 1, 185),
 (11, 1, 195),
 (11, 1, 197),
 (11, 1, 211),
 (11, 1, 214),
 (11, 1, 243),
 (11, 1, 318),
 (12, 1, 10),
 (12, 1, 54),
 (12, 1, 118),
 (12, 1, 215),
 (13, 1, 206),
 (14, 1, 42),
 (15, 1, 65),
 (15, 1, 120),
 (15, 1, 241),
 (15, 1, 260),
 (16, 2, 4),
 (16, 1, 5),
 (16, 1, 9),
 (16, 1, 10),
 (16, 1, 70),
 (16, 1, 145),
 (16, 1, 178),
 (16, 1, 321),
 (17, 2, 7),
 (17, 1, 115),
 (17, 1, 131),
 (17, 1, 162),
 (17, 1, 179),
 (17, 1, 185),
 (17, 1, 195),
 (17, 1, 203),
 (17, 1, 206),
 (17, 1, 207),
 (18, 1, 82),
 (18, 1, 127),
 (18, 1, 141),
 (18, 1, 144),
 (18, 1, 221),
 (19, 1, 11),
 (19, 1, 103),
 (19, 1, 146),
 (19, 1, 203),
 (19, 1, 214),
 (19, 1, 216),
 (19, 1, 321),
 (20, 1, 7),
 (20, 1, 11),
 (20, 1, 30),
 (20, 1, 72),
 (20, 1, 88),
 (20, 1, 252),
 (20, 2, 257),
 (20, 1, 260),
 (20, 1, 267),
 (21, 1, 1),
 (21, 1, 116),
 (21, 1, 118),
 (21, 1, 119),
 (21, 1, 121),
 (21, 1, 141),
 (21, 2, 207),
 (21, 1, 235),
 (22, 1, 33),
 (22, 1, 80),
 (23, 10, 1),
 (24, 12, 90),
 (25, 1, 5),
 (25, 1, 6),
 (25, 1, 75),
 (25, 1, 91),
 (25, 1, 119),
 (25, 1, 147),
 (25, 1, 180),
 (25, 1, 241),
 (26, 1, 10),
 (26, 1, 216),
 (27, 1, 101),
 (27, 1, 117),
 (27, 1, 121),
 (27, 2, 183),
 (27, 1, 238),
 (27, 1, 246),
 (27, 1, 305),
 (28, 1, 5),
 (28, 1, 74),
 (28, 1, 321),
 (29, 1, 39),
 (29, 1, 82),
 (29, 1, 132),
 (29, 1, 195),
 (29, 1, 200),
 (29, 1, 261),
 (30, 1, 5),
 (30, 1, 73),
 (30, 1, 79),
 (30, 1, 89),
 (30, 1, 164),
 (30, 1, 191),
 (30, 1, 195),
 (30, 1, 196),
 (30, 1, 246),
 (30, 2, 321),
 (31, 20, 221),
 (32, 13, 6),
 (33, 1, 206),
 (33, 1, 218),
 (33, 1, 221),
 (33, 1, 258),
 (34, 1, 49),
 (35, 1, 1),
 (35, 1, 9),
 (35, 1, 67),
 (35, 1, 194),
 (35, 1, 322),
 (36, 1, 16),
 (36, 1, 77),
 (36, 1, 91),
 (36, 1, 120),
 (36, 1, 127),
 (36, 1, 183),
 (36, 1, 215),
 (37, 1, 55),
 (37, 1, 81),
 (37, 1, 91),
 (37, 1, 115),
 (37, 1, 145),
 (37, 1, 257),
 (37, 1, 259),
 (37, 1, 260),
 (37, 1, 271),
 (37, 1, 321),
 (38, 1, 16),
 (38, 1, 119),
 (38, 1, 126),
 (38, 1, 207),
 (38, 1, 234),
 (38, 1, 255),
 (38, 1, 278),
 (39, 1, 19),
 (39, 1, 138),
 (39, 1, 141),
 (39, 1, 189),
 (39, 1, 196),
 (39, 1, 203),
 (39, 1, 269),
 (39, 1, 293),
 (40, 1, 11),
 (40, 1, 44),
 (40, 1, 54),
 (40, 1, 132),
 (40, 1, 183),
 (40, 1, 194),
 (40, 1, 207),
 (40, 1, 215),
 (40, 1, 221),
 (40, 1, 321),
 (41, 1, 16),
 (41, 1, 61),
 (41, 1, 160),
 (41, 1, 203),
 (41, 1, 218),
 (41, 1, 233),
 (41, 1, 260),
 (41, 1, 321),
 (42, 1, 1),
 (42, 1, 16),
 (42, 1, 46),
 (42, 2, 55),
 (42, 1, 72),
 (42, 1, 83),
 (42, 1, 214),
 (42, 1, 243),
 (43, 1, 73),
 (43, 1, 91),
 (43, 1, 135),
 (43, 1, 145),
 (43, 1, 163),
 (43, 1, 191),
 (43, 1, 214),
 (44, 12, 242),
 (45, 1, 11),
 (45, 1, 17),
 (45, 1, 146),
 (45, 1, 150),
 (45, 1, 155),
 (45, 1, 162),
 (45, 1, 214),
 (45, 1, 221),
 (46, 11, 214),
 (47, 1, 1),
 (47, 1, 37),
 (47, 1, 39),
 (47, 1, 171),
 (47, 1, 203),
 (47, 1, 271),
 (48, 1, 7),
 (48, 1, 36),
 (48, 1, 91),
 (48, 1, 182),
 (49, 2, 15),
 (49, 1, 91),
 (49, 1, 154),
 (49, 1, 207),
 (50, 1, 122),
 (50, 1, 147),
 (50, 1, 157),
 (50, 1, 178),
 (50, 1, 199),
 (50, 1, 270),
 (50, 1, 321),
 (51, 1, 5),
 (51, 1, 13),
 (51, 1, 39),
 (51, 1, 54),
 (51, 1, 92),
 (51, 1, 1

In [8]:
%%sql 
WITH purchases AS 
(SELECT il.invoice_id, COUNT(t.track_id) number_tracks, t.album_id 
FROM invoice_line il 
INNER JOIN track t on t.track_id = il.track_id
GROUP BY il.invoice_id, t.album_id),

album_list AS 
(SELECT DISTINCT(t.album_id), a.title, COUNT(t.track_id) album_tracks
FROM track t
LEFT JOIN album a ON a.album_id = t.album_id
GROUP BY a.album_id),

full_purchases AS
(SELECT p.invoice_id, al.album_id, al.album_tracks, p.number_tracks, CASE WHEN (al.album_tracks = p.number_tracks) THEN 1 ELSE 0
 END AS full_album_sale
 FROM purchases p
 LEFT JOIN album_list al ON al.album_id = p.album_id
 GROUP BY p.invoice_id)

SELECT COUNT(DISTINCT(invoice_id)) total_sales,
SUM(full_album_sale) number_full_album_sales,
COUNT(full_album_sale)-SUM(full_album_sale) track_sales,
(SUM(full_album_sale)/CAST(COUNT(invoice_id) AS float))*100 percentage_full_album_sales,
((COUNT(full_album_sale)-SUM(full_album_sale))/CAST(COUNT(invoice_id) AS float))*100 percentage_track_sales
FROM full_purchases


 * sqlite:///chinook.db
Done.


[(614, 117, 497, 19.055374592833875, 80.94462540716613)]

Out of 614 sales 184 were full album sales and 430 purchases of one or more tracks but not a complete album. Almost 30% of all purchases are full album purchases. While this is not the majority of purchases, it still is a significant portion. It is also a very lucrative portion, because those purchase usually contain 10 or more tracks, while the non-album purchases are often a single track.